File: src/main/kotlin/sunsetsatellite/lang/sunlite/TypeCollector.kt
package sunsetsatellite.lang.sunlite

import sunsetsatellite.vm.sunlite.SLNativeFuncObj
import sunsetsatellite.vm.sunlite.VM

class TypeCollector(val sunlite: Sunlite): Stmt.Visitor<Unit>, Expr.Visitor<Unit> {

    var currentClass: Stmt.Class? = null
    var currentScopeCandidate: Scope? = null

    abstract inner class ElementPrototype {
        abstract fun getElementType(): Type
        abstract fun isConstant(): Boolean
    }

    inner class VariablePrototype(val type: Type, val constant: Boolean): ElementPrototype() {
        override fun toString(): String {
            return ": $type"
        }

        override fun getElementType(): Type {
            return type
        }

        override fun isConstant(): Boolean {
            return constant
        }
    }

    inner class FunctionPrototype(
        val name: Token,
        val params: List<Param>,
        val returnType: Type,
        val typeParams: List<Param> = listOf()
    ): ElementPrototype(){
        override fun toString(): String {
            return "(${params.joinToString()}): $returnType"
        }

        override fun getElementType(): Type {
            if(typeParams.isNotEmpty()){
                return Type.ofGenericFunction(name.lexeme, returnType, params, typeParams)
            }
            return Type.ofFunction(name.lexeme, returnType, params)
        }

        override fun isConstant(): Boolean {
            return true
        }
    }

    class Scope(val name: Token, val contents: MutableMap<Token, ElementPrototype>, val depth: Int = -1, var outer: Scope? = null, var inner: MutableList<Scope> = mutableListOf()) {
        override fun toString(): String {
            return "scope '${name.lexeme}'"
        }
    }

    val typeHierarchy: MutableMap<String, Triple<String,List<String>,List<String>>> = mutableMapOf()
    val typeScopes: MutableList<Scope> = mutableListOf()
    var path: String? = null
    var currentScope: Scope? = Scope(Token.identifier("<global>",-1,"<global>"), mutableMapOf())

    init {
        typeScopes.add(currentScope!!)
        VM.globals.filter { it.value is SLNativeFuncObj }.forEach {
            addFunction(Token.identifier(it.key,-1,"<global>"),listOf(), (it.value as SLNativeFuncObj).value.returnType)
        }
    }

    fun addVariable(name: Token, type: Type, constant: Boolean = false) {
        if(currentScope?.contents?.mapKeys { it.key.lexeme }?.containsKey(name.lexeme) == true) sunlite.error(name, "Variable '${name.lexeme}' already declared in this scope.")
        currentScope?.contents?.put(name, VariablePrototype(type, constant))
    }
    fun addFunction(name: Token, params: List<Param>, returnType: Type, typeParams: List<Param> = listOf()) {
        if(typeScopes[0].contents.mapKeys { it.key.lexeme }.containsKey(name.lexeme)) sunlite.error(name, "Cannot overwrite global function '${name.lexeme}'.")
        if(currentScope?.contents?.mapKeys { it.key.lexeme }?.containsKey(name.lexeme) == true) sunlite.error(name, "Function '${name.lexeme}' already declared in this scope.")
        currentScope?.contents?.put(name, FunctionPrototype(name, params, returnType, typeParams))
    }

    fun addScope(name: Token){
        if(currentScope == null){
            currentScope = Scope(name, mutableMapOf())
            typeScopes.add(currentScope!!)
        } else {
            val scope = Scope(name, mutableMapOf(), currentScope!!.depth+1)
            currentScope!!.inner.add(scope)
            scope.outer = currentScope
            currentScope = scope
        }
    }

    fun removeScope(){
        currentScope = currentScope?.outer
    }

    fun collect(statements: List<Stmt>,path: String? = null){
        this.path = path
        addScope(Token.identifier(path.toString(),-1,path))
        statements.forEach { it.accept(this) }
        removeScope()
    }

    fun getValidScope(scope: Scope?, name: Token, enclosing: Token? = null, offset: Int = 0, enclosingScope: Scope? = null): Scope? {
        var enclosingS: Scope? = enclosingScope
        if(scope == null) return null
        if(scope.name.lexeme == enclosing?.lexeme){
            enclosingS = scope
        }
        if(scope.contents.keys.map { it.lexeme }.contains(name.lexeme)){
            currentScopeCandidate = scope
        }
        if(enclosingS != null && scope.depth <= enclosingS.depth + offset && scope.contents.keys.map { it.lexeme }.contains(name.lexeme)){
            return scope
        } else {
            var valid: Scope? = null
            scope.inner.forEach {
                valid = getValidScope(it, name, enclosing, offset, enclosingS)
                if(valid != null) return valid
            }
            return valid
        }
    }

    fun findType(name: Token, enclosing: Token? = null, offset: Int = 0): ElementPrototype? {
        currentScopeCandidate = null
        val globalScope = typeScopes[0]
        if(globalScope.contents.keys.map { it.lexeme }.contains(name.lexeme)){
            return globalScope.contents.mapKeys { it.key.lexeme }[name.lexeme]
        }
        val scope = getValidScope(typeScopes.firstOrNull(), name, enclosing, offset)
        if(scope == null && currentScopeCandidate == null) return null
        if(currentScopeCandidate != null) return currentScopeCandidate!!.contents.mapKeys { it.key.lexeme }[name.lexeme]
        return scope?.contents?.mapKeys { it.key.lexeme }[name.lexeme]
    }

    override fun visitExprStmt(stmt: Stmt.Expression) {
        stmt.expr.accept(this)
    }

    override fun visitPrintStmt(stmt: Stmt.Print) {
        // nothing to collect
    }

    override fun visitVarStmt(stmt: Stmt.Var) {
        addVariable(stmt.name, stmt.type, stmt.modifier == FieldModifier.CONST || stmt.modifier == FieldModifier.STATIC_CONST)
        stmt.initializer?.accept(this)
    }

    override fun visitBlockStmt(stmt: Stmt.Block) {
        addScope(Token.identifier("<block>",stmt.lineNumber,stmt.currentFile))
        stmt.statements.forEach { it.accept(this) }
        removeScope()
    }

    override fun visitIfStmt(stmt: Stmt.If) {
        stmt.condition.accept(this)
        stmt.thenBranch.accept(this)
        stmt.elseBranch?.accept(this)
    }

    override fun visitWhileStmt(stmt: Stmt.While) {
        stmt.condition.accept(this)
        stmt.body.accept(this)
    }

    override fun visitBreakStmt(stmt: Stmt.Break) {
        // nothing to collect
    }

    override fun visitContinueStmt(stmt: Stmt.Continue) {
        // nothing to collect
    }

    override fun visitFunctionStmt(stmt: Stmt.Function) {
        val typeParams = stmt.typeParams.toMutableList()
        currentClass?.typeParameters?.forEach { typeParams.add(it) }
        addFunction(stmt.name, stmt.params, stmt.returnType, typeParams)
        addScope(stmt.name)
        stmt.params.forEach { addVariable(it.token, it.type) }
        stmt.body.forEach { it.accept(this) }
        removeScope()
    }

    override fun visitReturnStmt(stmt: Stmt.Return) {
        stmt.value?.accept(this)
    }

    override fun visitClassStmt(stmt: Stmt.Class) {
        val classParams = stmt.methods.find { it.name.lexeme == "init" }?.params ?: listOf()
        addVariable(stmt.name, Type.ofClass(stmt.name.lexeme, classParams))
        addScope(stmt.name)
        currentClass = stmt
        val superclass = stmt.superclass?.name?.lexeme ?: "<nil>"
        val superinterfaces = stmt.superinterfaces.map { it.name.lexeme }.toMutableList()
        if(typeHierarchy.containsKey(stmt.name.lexeme)) sunlite.error(stmt.name, "Class '${stmt.name.lexeme}' already defined.")
        typeHierarchy[stmt.name.lexeme] = Triple(superclass, superinterfaces, stmt.typeParameters.map { it.token.lexeme })
        stmt.superclass?.let {
            addVariable(Token.identifier("<superclass>",it.getLine(),it.getFile()),Type.ofClass(it.name.lexeme))
        }
        stmt.superinterfaces.forEach {
            addVariable(Token.identifier("<superinterface ${it.name.lexeme}>",it.getLine(),it.getFile()),Type.ofClass(it.name.lexeme))
        }
        stmt.fieldDefaults.forEach { it.accept(this) }
        stmt.methods.forEach { it.accept(this) }
        stmt.typeParameters.forEach { addVariable(Token.identifier("<${it.token.lexeme}>",it.token.line, it.token.file), Type.Parameter(it.token)) }
        currentClass = null
        removeScope()
    }

    override fun visitInterfaceStmt(stmt: Stmt.Interface) {
        addVariable(stmt.name, Type.ofClass(stmt.name.lexeme))
        val superinterfaces = stmt.superinterfaces.map { it.name.lexeme }.toMutableList()
        if(typeHierarchy.containsKey(stmt.name.lexeme)) sunlite.error(stmt.name, "Class '${stmt.name.lexeme}' already defined.")
        typeHierarchy[stmt.name.lexeme] = Triple("<nil>",superinterfaces,stmt.typeParameters.map { it.token.lexeme })
        addScope(stmt.name)
        stmt.methods.forEach { it.accept(this) }
        removeScope()
    }

    override fun visitImportStmt(stmt: Stmt.Import) {
        // nothing to collect
    }

    override fun visitTryCatchStmt(stmt: Stmt.TryCatch) {
        stmt.tryBody.accept(this)
        stmt.catchBody.accept(this)
    }

    override fun visitThrowStmt(stmt: Stmt.Throw) {
        stmt.expr.accept(this)
    }

    override fun visitBinaryExpr(expr: Expr.Binary) {
        expr.left.accept(this)
        expr.right.accept(this)
    }

    override fun visitGroupingExpr(expr: Expr.Grouping) {
        expr.expression.accept(this)
    }

    override fun visitUnaryExpr(expr: Expr.Unary) {
        expr.right.accept(this)
    }

    override fun visitLiteralExpr(expr: Expr.Literal) {

    }

    override fun visitVariableExpr(expr: Expr.Variable) {

    }

    override fun visitAssignExpr(expr: Expr.Assign) {
        expr.value.accept(this)
    }

    override fun visitLogicalExpr(expr: Expr.Logical) {
        expr.left.accept(this)
        expr.right.accept(this)
    }

    override fun visitCallExpr(expr: Expr.Call) {
        expr.callee.accept(this)
        expr.arguments.forEach { it.accept(this) }
    }

    override fun visitLambdaExpr(expr: Expr.Lambda) {
        expr.function.accept(this);
    }

    override fun visitGetExpr(expr: Expr.Get) {
        expr.obj.accept(this)
    }

    override fun visitArrayGetExpr(expr: Expr.ArrayGet) {
        expr.obj.accept(this)
        expr.what.accept(this)
    }

    override fun visitArraySetExpr(expr: Expr.ArraySet) {
        expr.obj.accept(this)
        expr.what.accept(this)
        expr.value.accept(this)
    }

    override fun visitSetExpr(expr: Expr.Set) {
        expr.obj.accept(this)
        expr.value.accept(this)
    }

    override fun visitThisExpr(expr: Expr.This) {

    }

    override fun visitSuperExpr(expr: Expr.Super) {

    }

    override fun visitCheckExpr(expr: Expr.Check) {
        expr.left.accept(this)
    }

    override fun visitCastExpr(expr: Expr.Cast) {
        expr.left.accept(this)
    }

}

File: src/main/csharp/SunliteSharp/Core/Compiler/TypeCollector.cs
using SunliteSharp.Core.AST;
using SunliteSharp.Core.Modifier;
using SunliteSharp.Runtime;

namespace SunliteSharp.Core.Compiler;

public class TypeCollector : Stmt.Visitor, Expr.Visitor
{
    
    private readonly Sunlite Sl;

    public TypeCollector(Sunlite sunlite)
    {
        Sl = sunlite;
        TypeScopes.Add(CurrentScope);
        foreach (KeyValuePair<string, AnySLValue> kvp in Sl.Vm!.Globals.Where(kv => kv.Value is SLNativeFuncObj))
        {
            var token = Token.Identifier(kvp.Key, -1, "<global>");
            var nativeFunc = (SLNativeFuncObj)kvp.Value;
            AddFunction(token, [], nativeFunc.Value.ReturnType);
        }
    }
    
    public interface IElementPrototype
    {
        public Type GetElementType();
        public bool IsConstant();
    }

    public record VariablePrototype(Type Type, bool Constant) : IElementPrototype
    {
        public override string ToString()
        {
            return $": {Type}";
        }

        public Type GetElementType() => Type;
        
        public bool IsConstant() => Constant;
    }
    
    public record FunctionPrototype(
        Token Name,
        List<Param> Params,
        Type ReturnType,
        List<Param> TypeParams
    ) : IElementPrototype
    {
        public override string ToString() => $"({string.Join(", ", Params)}): {ReturnType}";

        public Type GetElementType() 
            => TypeParams.Count > 0
                ? Type.OfGenericFunction(Name.Lexeme, ReturnType, Params, TypeParams)
                : Type.OfFunction(Name.Lexeme, ReturnType, Params);

        public bool IsConstant() => false;
    }

    public record Scope(
        Token Name,
        Dictionary<Token, IElementPrototype> Contents,
        int Depth = -1
    )
    {
        public Scope? Outer;
        public List<Scope> Inner = [];

        public override string ToString() => $"scope '{Name.Lexeme}'";
    }
    
    public Dictionary<string, (string, List<string>, List<string>)> TypeHierarchy = new();
    public List<Scope> TypeScopes = [];
    public string CurrentFile = "<unknown>";
    
    public Scope? CurrentScope = new Scope(Token.Identifier("<global>", "<global>"), []);
    public Stmt.Class? CurrentClass = null;
    public Scope? CurrentScopeCandidate = null;
    
    public void AddVariable(Token name, Type type, bool constant = false)
    {
        if (CurrentScope?.Contents.Keys.Any(k => k.Lexeme == name.Lexeme) == true)
            Sl.Error(name, $"Variable '{name.Lexeme}' already declared in this scope.");

        if (CurrentScope is not null) CurrentScope.Contents[name] = new VariablePrototype(type, constant);
    }

    public void AddFunction(Token name, List<Param> parameters, Type returnType, List<Param>? typeParams = null)
    {
        typeParams ??= [];

        if (CurrentScope?.Contents.Keys.Any(k => k.Lexeme == name.Lexeme) == true)
            Sl.Error(name, $"Function '{name.Lexeme}' already declared in this scope.");

        if (CurrentScope is not null) CurrentScope.Contents[name] = new FunctionPrototype(name, parameters, returnType, typeParams);
    }

    public void AddScope(Token name)
    {
        if (CurrentScope is null)
        {
            CurrentScope = new Scope(name, []);
            TypeScopes.Add(CurrentScope);
        }
        else
        {
            var scope = new Scope(name, [], CurrentScope.Depth + 1)
            {
                Outer = CurrentScope
            };
            CurrentScope.Inner.Add(scope);
            CurrentScope = scope;
        }
    }

    public void RemoveScope()
    {
        CurrentScope = CurrentScope?.Outer;
    }

    public void Collect(List<Stmt> statements, string filePath)
    {
        CurrentFile = filePath;
        AddScope(Token.Identifier(filePath,filePath));
        foreach (var statement in statements)
        {
            statement.Accept(this);
        }
        RemoveScope();
    }

    public Scope? GetValidScope(
        Scope? scope,
        Token name,
        Token? enclosing = null,
        int offset = 0,
        Scope? enclosingScope = null)
    {
        var enclosingS = enclosingScope;
        if (scope == null) return null;

        if (scope.Name.Lexeme == enclosing?.Lexeme)
        {
            enclosingS = scope;
        }

        if (scope.Contents.Keys.Any(k => k.Lexeme == name.Lexeme))
        {
            CurrentScopeCandidate = scope;
        }

        if (enclosingS is not null &&
            scope.Depth <= enclosingS.Depth + offset &&
            scope.Contents.Keys.Any(k => k.Lexeme == name.Lexeme))
        {
            return scope;
        }
        else
        {
            foreach (var innerScope in scope.Inner)
            {
                var valid = GetValidScope(innerScope, name, enclosing, offset, enclosingS);
                if (valid is not null) return valid;
            }
            return null;
        }
    }

    public IElementPrototype? FindType(Token name, Token? enclosing = null, int offset = 0)
    {
        CurrentScopeCandidate = null;
        var globalScope = TypeScopes[0];

        if (globalScope.Contents.Keys.Any(k => k.Lexeme == name.Lexeme))
        {
            return globalScope.Contents.First(kv => kv.Key.Lexeme == name.Lexeme).Value;
        }

        var scope = GetValidScope(TypeScopes.FirstOrDefault(), name, enclosing, offset);
        if (scope == null && CurrentScopeCandidate == null) return null;

        if (CurrentScopeCandidate != null)
        {
            return CurrentScopeCandidate.Contents.First(kv => kv.Key.Lexeme == name.Lexeme).Value;
        }

        return scope?.Contents.First(kv => kv.Key.Lexeme == name.Lexeme).Value;
    }


    public void VisitExprStmt(Stmt.Expression stmt)
    {
        stmt.Expr.Accept(this);
    }
    
    public void VisitVarStmt(Stmt.Var stmt)
    {
        AddVariable(stmt.Name, stmt.Type, stmt.Modifier is FieldModifier.Const or FieldModifier.StaticConst);
        stmt.Initializer?.Accept(this);
    }

    public void VisitBlockStmt(Stmt.Block stmt)
    {
        AddScope(Token.Identifier("<block>", stmt.LineNumber, stmt.CurrentFile));
        foreach (var statement in stmt.Statements)
        {
            statement.Accept(this);
        }
        RemoveScope();
    }

    public void VisitIfStmt(Stmt.If stmt)
    {
        stmt.Condition.Accept(this);
        stmt.ThenBranch.Accept(this);
        stmt.ElseBranch?.Accept(this);
    }

    public void VisitWhileStmt(Stmt.While stmt)
    {
        stmt.Condition.Accept(this);
        stmt.Body.Accept(this);
    }

    public void VisitBreakStmt(Stmt.Break stmt)
    {
        // nothing to collect
    }

    public void VisitContinueStmt(Stmt.Continue stmt)
    {
        // nothing to collect
    }

    public void VisitFunctionStmt(Stmt.Function stmt)
    {
        var typeParams = new List<Param>(stmt.TypeParams);
        if (CurrentClass != null)
        {
            typeParams.AddRange(CurrentClass.TypeParams);
        }

        AddFunction(stmt.Name, stmt.Params, stmt.ReturnType, typeParams);
        AddScope(stmt.Name);
        foreach (var param in stmt.Params)
        {
            AddVariable(param.Token, param.Type);
        }
        foreach (var bodyStmt in stmt.Body)
        {
            bodyStmt.Accept(this);
        }
        RemoveScope();
    }

    public void VisitReturnStmt(Stmt.Return stmt)
    {
        stmt.Value?.Accept(this);
    }

    public void VisitClassStmt(Stmt.Class stmt)
    {
        List<Param> classParams = stmt.Methods.FirstOrDefault(m => m.Name.Lexeme == "init")?.Params ?? new List<Param>();
        AddVariable(stmt.Name, Type.OfClass(stmt.Name.Lexeme, classParams));
        AddScope(stmt.Name);
        CurrentClass = stmt;

        var superclass = stmt.Superclass?.Name.Lexeme ?? "<nil>";
        List<string> superinterfaces = stmt.Superinterfaces.Select(i => i.Name.Lexeme).ToList();

        if (TypeHierarchy.ContainsKey(stmt.Name.Lexeme))
        {
            Sl.Error(stmt.Name, $"Class '{stmt.Name.Lexeme}' already defined.");
        }
        TypeHierarchy[stmt.Name.Lexeme] = (superclass, superinterfaces, stmt.TypeParams.Select(tp => tp.Token.Lexeme).ToList());

        if (stmt.Superclass != null)
        {
            AddVariable(Token.Identifier("<superclass>", stmt.Superclass.GetLine(), stmt.Superclass.GetFile()), Type.OfClass(stmt.Superclass.Name.Lexeme,[]));
        }

        foreach (var superinterface in stmt.Superinterfaces)
        {
            AddVariable(Token.Identifier($"<superinterface {superinterface.Name.Lexeme}>", superinterface.GetLine(), superinterface.GetFile()), Type.OfClass(superinterface.Name.Lexeme,[]));
        }

        foreach (var fieldDefault in stmt.FieldDefaults)
        {
            fieldDefault.Accept(this);
        }
        foreach (var method in stmt.Methods)
        {
            method.Accept(this);
        }
        foreach (var typeParam in stmt.TypeParams)
        {
            AddVariable(Token.Identifier($"<{typeParam.Token.Lexeme}>", typeParam.Token.Line, typeParam.Token.File), new Type.Parameter(typeParam.Token));
        }
        CurrentClass = null;
        RemoveScope();
    }

    public void VisitInterfaceStmt(Stmt.Interface stmt)
    {
        AddVariable(stmt.Name, Type.OfClass(stmt.Name.Lexeme,[]));
        List<string> superinterfaces = stmt.Superinterfaces.Select(i => i.Name.Lexeme).ToList();

        if (TypeHierarchy.ContainsKey(stmt.Name.Lexeme))
        {
            Sl.Error(stmt.Name, $"Class '{stmt.Name.Lexeme}' already defined.");
        }
        TypeHierarchy[stmt.Name.Lexeme] = ("<nil>", superinterfaces, stmt.TypeParameters.Select(tp => tp.Token.Lexeme).ToList());

        AddScope(stmt.Name);
        foreach (var method in stmt.Methods)
        {
            method.Accept(this);
        }
        RemoveScope();
    }

    public void VisitImportStmt(Stmt.Import stmt)
    {
        // nothing to collect
    }

    public void VisitTryCatchStmt(Stmt.TryCatch stmt)
    {
        stmt.TryBody.Accept(this);
        stmt.CatchBody.Accept(this);
    }

    public void VisitThrowStmt(Stmt.Throw stmt)
    {
        stmt.Expr.Accept(this);
    }

    public void VisitBinaryExpr(Expr.Binary expr)
    {
        expr.Left.Accept(this);
        expr.Right.Accept(this);
    }

    public void VisitGroupingExpr(Expr.Grouping expr)
    {
        expr.Expression.Accept(this);
    }

    public void VisitUnaryExpr(Expr.Unary expr)
    {
        expr.Right.Accept(this);
    }

    public void VisitLiteralExpr(Expr.Literal expr)
    {
        // nothing to collect
    }

    public void VisitVariableExpr(Expr.Variable expr)
    {
        // nothing to collect
    }

    public void VisitAssignExpr(Expr.Assign expr)
    {
        expr.Value.Accept(this);
    }

    public void VisitLogicalExpr(Expr.Logical expr)
    {
        expr.Left.Accept(this);
        expr.Right.Accept(this);
    }

    public void VisitCallExpr(Expr.Call expr)
    {
        expr.Callee.Accept(this);
        foreach (var argument in expr.Arguments)
        {
            argument.Accept(this);
        }
    }

    public void VisitLambdaExpr(Expr.Lambda expr)
    {
        expr.Function.Accept(this);
    }

    public void VisitGetExpr(Expr.Get expr)
    {
        expr.Obj.Accept(this);
    }

    public void VisitArrayGetExpr(Expr.ArrayGet expr)
    {
        expr.Obj.Accept(this);
        expr.What.Accept(this);
    }

    public void VisitArraySetExpr(Expr.ArraySet expr)
    {
        expr.Obj.Accept(this);
        expr.What.Accept(this);
        expr.Value.Accept(this);
    }

    public void VisitSetExpr(Expr.Set expr)
    {
        expr.Obj.Accept(this);
        expr.Value.Accept(this);
    }

    public void VisitThisExpr(Expr.This expr)
    {
        // nothing to collect
    }

    public void VisitSuperExpr(Expr.Super expr)
    {
        // nothing to collect
    }

    public void VisitCheckExpr(Expr.Check expr)
    {
        expr.Left.Accept(this);
    }

    public void VisitCastExpr(Expr.Cast expr)
    {
        expr.Left.Accept(this);
    }

}